include/circt/Dialect/Datapath/DatapathOps.h

```diff
@@ -17,4 +17,86 @@
 #define GET_OP_CLASSES
 #include "circt/Dialect/Datapath/Datapath.h.inc"
 
+namespace circt {
+namespace datapath {
+
+// A pair used to track compressor bits
+struct CompressorBit {
+  Value val;
+  size_t delay;
+};
+
+// A general compressor tree interface based around a column-wise
+// representation. Contains alternative compression algorithms.
+class CompressorTree {
+public:
+  // Constructor takes addends as input and converts to column representation
+  CompressorTree(const SmallVector<SmallVector<Value>> &addends, Location loc);
+
+  // Get the number of columns (bit positions)
+  size_t getWidth() const { return columns.size(); }
+
+  // Get the maximum height of the addend array
+  size_t getMaxHeight() const;
+
+  // Get the target height of next stage
+  size_t getNextStageTargetHeight() const;
+
+  // Update the input delays based on longest path analysis
+  void withInputDelays(const SmallVector<SmallVector<int64_t>> inputDelays);
+
+  // Apply a compression step (reduce columns with >2 bits using compressors)
+  SmallVector<Value> compressToHeight(OpBuilder &builder, size_t targetHeight);
+
+  // Debug: print the tree structure
+  void dump() const;
+
+private:
+  // Original addends representation as bitvectors (kept for reference)
+  SmallVector<SmallVector<Value>> originalAddends;
+
+  // Column-wise bit storage - columns[i] contains all bits at bit position i
+  SmallVector<SmallVector<CompressorBit>> columns;
+
+  // Whether to use a timing driven compression algorithm
+  // If true, use a timing driven compression algorithm.
+  // If false, use a simple compression algorithm (e.g., Wallace).
+  bool usingTiming;
+
+  // Bitwidth of compressor tree
+  size_t width;
```

Comment:
Can you pass width explicitly and make it const?
```suggestion
  const size_t width;
```